zcash_client_backend/proptest-regressions/zip321.txt

```diff
@@ -0,0 +1,8 @@
+# Seeds for failure cases proptest has generated in the past. It is
+# automatically read and these particular cases re-run before any
+# novel cases are generated.
+#
+# It is recommended to check this file in to source control so that
+# everyone who runs the test benefits from these saved cases.
+cc 150e79b7e542f49689fad3c6e2bb3f830a78e4f2035e9fbab094d62e65f361d5 # shrinks to req = Zip321Request { payments: [Zip321Payment { recipient_address: SaplingAddress(PaymentAddress { pk_d: SubgroupPoint(ExtendedPoint { u: 0x25621e6b6515e85443822759660e9eef04918451ac4cd1263b8087d00689fdff, v: 0x11baaa27bc00dd67e8e03c98ca86c14b1f3ac420242f8e5cc5f9db1dced59b64, z: 0x168e654728395ce631656f756b4ae31dd19b6be3a31f55fd801f8272c5eb1dac, t1: 0x312ff1fd7ffb718f705f340bee2d80a6a0182efd81dd8f2a78cc468320be470f, t2: 0x0af52c1864feef43acb492b8bb106570ecca0ecd73567d8b2888769bb6022564 }), diversifier: Diversifier([59, 246, 250, 31, 131, 191, 69, 99, 200, 167, 19]) }), amount: Amount(0), memo: None, label: None, message: Some(""), other_params: [] }] }
```

Comment:
Is there a best practice around formatting this kind of test vector?
This is my first look at `proptest` and it seems like making this human readable would be helpful.
I just checked:  https://github.com/zcash/librustzcash/pull/294/commits/ff1ecaf950cceea8d83a536226a8667af15c7e58

It seems like this comment is still relevant.

Reply:
I think I will tinker a bit with formats...

Reply:
These are automatically generated by proptest, so it probably shouldn't be messed with.

Reply:
I should elaborate: as proptest generates samples, when a sample that fails the assertions is generated it is persisted to this file and then automatically checked on subsequent proptest runs (using the value of `cc` as the seed for randomization). So this is never human-edited at all, just saved after proptest finds a failure.

Reply:
Awesome!
I'm really excited to learn about `proptest`, BTW.
A few thoughts:
(1)   It seems like if we had an automated formatter that was also checked in to git, so we could verifiably convert from "native" format to "human-readable" and back, then we could just "punch a button" to get the relevant format.  (It seems like it might not be worth the effort, out-of-the box, but I guess that's a function of how much effort it is to write (among other things!).)  On the other hand it might be a handy test-utility.
(2)  What if on every iteration of a proptest there was a _very_ low probability (tunable) that it would record its `seed`?    <--  This probability would be independent of whether the test succeeded-or-failed.   This would provide a uniform distribution of seeds over time.   Then what if there were a class of tests that were known to fail _sometimes_ where that "sometimes" was unquantified/estimated.    Then the distribution (frequency) of failures in the "sometimes" tests would provide quantitative data.  (Actually I'm not sure I have come up with a very good example.)
  For example, what if there was a test that fails whenever some "transaction/block"? isn't gossiped about enough in some period of time.   That test could fail for various reasons, e.g. someone tripped over the ethernet cable, or the block was malformed and other nodes won't acknowledge it, etc.  The point is that the amount of time until the test fails would show timing information like:
  90% of the time, the test passes, 10% of the time the test (by design! fails).
  So then one day someone asks:   How often is two minutes insufficient for a FOO to be gossiped about?   we'd have an answer!

OK, so maybe all of the above is already discussed elsewhere, or maybe I should move the thought elsewhere.